plugin/pulpcore/plugin/stages/association_stages.py

```diff
@@ -0,0 +1,150 @@
+import asyncio
+from collections import defaultdict
+
+from django.db.models import Q
+
+from pulpcore.plugin.models import ProgressBar
+
+from .api import BaseStage
+
+
+class ContentUnitAssociation(BaseStage):
+    """
+    A Stages API stage that associates content units with `new_version`.
+
+    This stage stores all content unit types and unit keys in memory before running. This is done to
+    compute the units already associated but not received from `in_q`. These units are passed via
+    `out_q` to the next stage as a :class:`django.db.models.query.QuerySet`.
+
+    This stage creates a ProgressBar named 'Associating Content' that counts the number of units
+    associated. Since it's a stream the total count isn't known until it's finished.
+
+    Args:
+        new_version (:class:`~pulpcore.plugin.models.RepositoryVersion`): The repo version this
+            stage associates content with.
+    """
+
+    def __init__(self, new_version, *args, **kwargs):
+        self.new_version = new_version
+        self.unit_keys_by_type = defaultdict(set)
+        for unit in self.new_version.content.all():
+            unit = unit.cast()
+            self.unit_keys_by_type[type(unit)].add(unit.natural_key())
+        super().__init__(*args, **kwargs)
```

Comment:
Same re: calling *super* as: [here](https://github.com/pulp/pulp/pull/3559/files#r207544584)
Same re: *args, kwargs* as: [here](https://github.com/pulp/pulp/pull/3559/files#r207543875)

Reply:
yup will fix in next push.